### PR TITLE
TYP: _concat_same_type method of EA

### PR DESCRIPTION
**Pull request by simonjayhawkins**
some overlap with #37816

### CODE REVIEW

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -1468,10 +1472,19 @@ def _get_combined_data(
             axis=1,
         )
     else:
-        left = cast(Union["DatetimeArray", "TimedeltaArray"], left)
-        right = cast(Union["DatetimeArray", "TimedeltaArray"], right)
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:1472]: the casting doesnt help?

**Comment by jreback** [comment]
can you merge master

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -37,6 +48,8 @@
 
 _extension_array_shared_docs: Dict[str, str] = dict()
 
+ExtensionArrayT = TypeVar("ExtensionArrayT", bound="ExtensionArray")
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py:51]: I might have asked before, but can you explain again why this is needed, and the `"ExtensionArray"` as string version doesn't work? 


(I think I mentioned this before, but IMO we should have better documentation about the specific patterns we are using in pandas in our contributing guides)

Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:51]: > (I think I mentioned this before, but IMO we should have better documentation about the specific patterns we are using in pandas in our contributing guides)

these are not specific patterns to pandas. https://mypy.readthedocs.io/en/stable/generics.html#generic-functions and https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self

Comment (simonjayhawkins) [reply pandas/core/arrays/base.py]: > I might have asked before, but can you explain again why this is needed, and the `"ExtensionArray"` as string version doesn't work?

I do recall explaining it in the past (obviously didn't do a good job) won't bother with a generic explanation again, and next time I come across a concrete example of why this is needed will cc you in.

File: pandas/core/arrays/interval.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/interval.py:1474]: the casts as they stand give

```
pandas\core\arrays\interval.py:1478: error: List item 0 has incompatible type "Union[DatetimeArray, TimedeltaArray]"; expected "DatetimeArray"  [list-item]
pandas\core\arrays\interval.py:1478: error: List item 0 has incompatible type "Union[DatetimeArray, TimedeltaArray]"; expected "TimedeltaArray"  [list-item]
pandas\core\arrays\interval.py:1478: error: List item 1 has incompatible type "Union[DatetimeArray, TimedeltaArray]"; expected "DatetimeArray"  [list-item]
pandas\core\arrays\interval.py:1478: error: List item 1 has incompatible type "Union[DatetimeArray, TimedeltaArray]"; expected "TimedeltaArray"  [list-item]
```
 and would need to be changed. I can't immediately see what guarentees we have that justify the casts.

the 'ignore' is a fix later properly. (I guess the function signature here maybe incorrect, but not checked) and will circle back to this after getting started on the other EA methods as a precursor to adding type annotations to #35259

If this is not a false positive, it needs more investigation before casting to silence the mypy errors.

**Comment by jreback** [comment]
thanks @simonjayhawkins

**Pull request merged**